pageserver/src/tenant/layer_map.rs

```diff
@@ -273,31 +289,46 @@ where
     ///
     /// Helper function for BatchedUpdates::insert_historic
     ///
-    pub(self) fn insert_historic_noflush(&mut self, layer: Arc<L>) {
+    /// TODO(chi): remove L generic so that we do not need to pass layer desc.
+    pub(self) fn insert_historic_noflush(
+        &mut self,
+        layer_desc: PersistentLayerDesc,
+        layer: Arc<L>,
+    ) {
+        self.mapping.insert(layer_desc.key(), layer.clone());
+
         // TODO: See #3869, resulting #4088, attempted fix and repro #4094
+
+        if Self::is_l0(&layer) {
+            self.l0_delta_layers.push(layer_desc.clone().into());
+        }
+
         self.historic.insert(
             historic_layer_coverage::LayerKey::from(&*layer),
-            Arc::clone(&layer),
+            layer_desc.into(),
         );
+    }
 
-        if Self::is_l0(&layer) {
-            self.l0_delta_layers.push(layer);
-        }
+    fn get_layer_from_mapping(&self, key: &PersistentLayerKey) -> &Arc<L> {
+        let layer = self.mapping.get(key).expect("inconsistent layer mapping");
+        layer
     }
 
     ///
     /// Remove an on-disk layer from the map.
     ///
     /// Helper function for BatchedUpdates::remove_historic
     ///
-    pub fn remove_historic_noflush(&mut self, layer: Arc<L>) {
+    pub fn remove_historic_noflush(&mut self, layer_desc: PersistentLayerDesc, layer: Arc<L>) {
         self.historic
             .remove(historic_layer_coverage::LayerKey::from(&*layer));
-
         if Self::is_l0(&layer) {
             let len_before = self.l0_delta_layers.len();
-            self.l0_delta_layers
-                .retain(|other| !Self::compare_arced_layers(other, &layer));
+            let mut l0_delta_layers = std::mem::take(&mut self.l0_delta_layers);
```

Comment:
Seems like the `std::mem::take` is a way to avoid `let other_layer = self.get_layer_from_mapping(&other.key()).clone()`. Isn't the latter cleaner? Yes, there's an `Arc::clone` but we avoid N queries to `self.get_layer_from_mapping`.

But maybe the bigger question here is "why was `l0_delta_layers` a `Vec`", and not a `HashMap`? Maybe because we didn't have a convenient layer key struct?

Reply:
It's the ownership thing, whereas we cannot hold l0_delta_layer mutable reference with self.get_layer_from_mapping (as an immutable reference). Therefore we have to first take it out, modify it, and then put it back.